plugins/logitech-tap/fu-logitech-tap-touch-device.c

```diff
@@ -0,0 +1,913 @@
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_tap_touch_device_attach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuLogitechTapTouchDevice *self = FU_LOGITECH_TAP_TOUCH_DEVICE(device);
+	g_autoptr(FuStructLogitechTapTouch) st = fu_struct_logitech_tap_touch_new();
+
+	/* sanity check */
+	if (!fu_device_has_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER)) {
+		g_debug("already in runtime mode, skipping");
+		return TRUE;
+	}
+
+	/* hid report to enable write and switch to application (AP) mode */
+	if (!fu_logitech_tap_touch_device_write_enable(self, FALSE, FALSE, 0, 0, error))
+		return FALSE;
+	fu_struct_logitech_tap_touch_set_payload_len(st, 0x01);
+	fu_struct_logitech_tap_touch_set_response_len(st, 0x0);
+	fu_byte_array_append_uint8(st, FU_STRUCT_LOGITECH_TAP_TOUCH_HID_CMD_SET_AP_MODE);
+	if (!fu_logitech_tap_touch_device_hid_transfer(self, st, 0, NULL, error))
+		return FALSE;
+
+	/* mode switch delay for application/bootloader */
+	fu_device_sleep(FU_DEVICE(self), 100);
+
+	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_tap_touch_device_probe(FuDevice *device, GError **error)
+{
+	/* FuUdevDevice->probe */
+	if (!FU_DEVICE_CLASS(fu_logitech_tap_touch_device_parent_class)->probe(device, error))
+		return FALSE;
+
+	/* ignore unsupported subsystems */
+	if (g_strcmp0(fu_udev_device_get_subsystem(FU_UDEV_DEVICE(device)), "hidraw") != 0) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "is not correct subsystem=%s, expected hidraw",
+			    fu_udev_device_get_subsystem(FU_UDEV_DEVICE(device)));
+		return FALSE;
+	}
+
+	/* set the physical ID */
+	return fu_udev_device_set_physical_id(FU_UDEV_DEVICE(device), "hid", error);
+}
+
+static gboolean
+fu_logitech_tap_touch_device_write_blocks(FuLogitechTapTouchDevice *self,
+					  const guint8 *data,
+					  guint datasz,
+					  guint32 start,
+					  guint32 end,
+					  guint16 firmware_checksum,
+					  gboolean in_ap,
+					  FuProgress *progress,
+					  GError **error)
+{
+	guint16 bufsize;
+	guint16 device_checksum = 0;
+
+	/* progress */
+	g_info("updating %s block. start: 0x%x, end:0x%x, checksum:0x%x",
+	       in_ap ? "AP" : "DF",
+	       start,
+	       end,
+	       firmware_checksum);
+
+	/* write */
+	fu_progress_set_id(progress, G_STRLOC);
+	/* integer round up */
+	fu_progress_set_steps(progress,
+			      (((end - start) + (TAP_TOUCH_TRANSFER_BLOCK_SIZE - 1)) /
+			       TAP_TOUCH_TRANSFER_BLOCK_SIZE));
+
+	/* hid report to enable write */
+	if (!fu_logitech_tap_touch_device_write_enable(self,
+						       FALSE,
+						       in_ap ? TRUE : FALSE,
+						       end + 1,
+						       firmware_checksum,
+						       error))
+		return FALSE;
+	fu_device_sleep(FU_DEVICE(self), 10);
+
+	// FIXME use FuChunk
+	for (guint32 i = start; i <= end; i += TAP_TOUCH_TRANSFER_BLOCK_SIZE) {
```

Comment:
Still this; I think FuChunk will clean up a lot of this

Reply:
Fixed